plugins/rrl/rrl.go

```diff
@@ -142,34 +141,32 @@ func (rrl *RRL) buildToken(rtype uint8, qtype uint16, name, remoteAddr string) s
 	return ""
 }
 
-// debit will decrement an existing response account in the rrl table by one and recalculate the current balance,
+// debit will update an existing response account in the rrl table and recalculate the current balance,
 // or if the response account does not exist, it will add it.
-func (rrl *RRL) debit(allowance float64, t string) (float64, error) {
+func (rrl *RRL) debit(allowance int64, t string) (int64, error) {
 	result := rrl.table.UpdateAdd(t,
-		// the 'update' function debits the account and returns the new balance
+		// the 'update' function updates the account and returns the new balance
 		func(el *interface{}) interface{} {
 			ra := (*el).(*ResponseAccount)
 			if ra == nil {
 				return nil
 			}
-			now := time.Now()
-			ra.balance += allowance*now.Sub(ra.lastCheck).Seconds() - 1
-			if ra.balance >= rrl.window {
+			now := time.Now().UnixNano()
+			balance := now - ra.allowTime - allowance
+			if balance >= rrl.window {
```

Comment:
got it. Should be `if balance >= second {` since I store all time values in nanoseconds

regarding balance scale - yes, internally, now balance is amount of time in nanoseconds during which we can send responses with defined rate. It can be converted to old representation (allowed number of responses) by dividing it by allowed interval, e.g. see my change in handler.go